File: eq.hs
module Eq where

import qualified Prelude

infixl 1 &&
(&&) :: Bool -> Bool -> Bool
True && True = True
_ && _ = False

data Bool = True | False
data Nat  = Zero | Suc Nat

class Eq α where
  eq :: α -> α -> Bool

instance Eq Nat where
  eq Zero    Zero    = True
  eq (Suc x) (Suc y) = eq x y
  eq _        _      = False

instance Eq α => Eq [α] where
  eq []       []       = True
  eq (x : xs) (y : ys) = eq x y && eq xs ys
  eq _           _     = False

isEq :: Bool
isEq = eq [Zero] [Zero]

File: eq.rs
use std::ops::BitAnd;

enum Bool { True, False }
enum Nat { Zero, Suc (Box<Nat>)  }

use Bool::*;
use Nat::*;

impl BitAnd for Bool {
    type Output = Bool;

    fn bitand(self, rhs: Self) -> Self::Output {
        match (self, rhs) {
            (True, True) => True,
            (_, _)       => False
        }
    }
}

trait Eq {
  fn eq(&self, rhs: &Self) -> Bool;
}

impl Eq for Nat {
  fn eq(&self, rhs: &Self) -> Bool {
    match (self, rhs) {
      (Zero  , Zero    ) => True,
      (Suc(x), Suc(y))   => x.eq(y),
      (_, _)        => False
    }
  }
}

impl<A: Eq> Eq for [A] {
  fn eq(&self, rhs: &Self) -> Bool {
    match (self, rhs) {
      ([], []) => True,
      ([x, xs @ ..], [y, ys @ ..]) 
        => x.eq(y) & xs.eq(ys),
      (_ , _)  => False
    }
  }
}

fn isEq() -> Bool { [Zero].eq(&[Zero]) }